### ISSUE TITLE
[python] Document all possible inputs and their support matrix.

### ISSUE THREAD

**Issue by trivialfis** [dmlc/xgboost#8541]
Working in progress summary for all supported Python inputs.

## Markers
- F: Not planned. (unless specifically requested and discussed for an actual use case)
- T: Supported.
- NE: Doesn't make sense to do so. For instance, `pd.Series` can not be multi-target label.
- NPA: Support with the help of numpy array.
- CPA: Support with the help of cupy array.
- SciCSR: Support with the help of scripy sparse CSR. The conversion to scipy CSR may or may not be possible. Raise a type error if conversion fails.
- H: Categorical feature with the help of `feature_types` parameter. (Except for dataframe, other data structures don't carry the feature type information like `category`, we use the extra parameter `feature_types` in DMatrix for specifying this information).
- FF: We can look forward to having its support in recent future if requested. For instance, pandas series(instead of 2d DF) as predictors for QDM.
- empty: To be filled in.

## Table Header
- `X` means predictor matrix.
- Meta info: label, weight, etc.
- Multi Label: 2-dim label for multi-target.
- Others: Anything else that we don't list here explicitly including formats like `lil`, `dia`, `bsr`. XGBoost will try to convert it into scipy csr.

## non-distributed data structures
| Name                    | DMatrix X | QuantileDMatrix X | Sklearn X | Meta Info | Categorical | Inplace prediction | Multi Label |
|:------------------------|:----------|-------------------|-----------|:----------|:------------|:-------------------|-------------|
| numpy.ndarray           | T         | T                 | T         | T         | H           | T                  | T           |
| scipy.sparse.csr        | T         | T                 | T         | NE        | H           | T                  | F           |
| scipy.sparse.csc        | T         | F                 | T         | NE        | H           | F                  | F           |
| scipy.sparse.coo        | SciCSR    | F                 | SciCSR    | NE        | H           | F                  | F           |
| uri                     | T         | F                 | F         | F         | H           | NE                 | F           |
| list                    | NPA       | NPA(2.0)          | T         | NPA       |             | NPA(2.0)           |             |
| tuple                   | NPA       | NPA(2.0)          | T         | NPA       | H           | NPA(2.0)           |             |
| pandas.DataFrame        | NPA       | NPA               | NPA       | NPA       | T           | NPA                | NPA         |
| pandas.Series           | NPA       | NPA(2.0)          |  T        | NPA       | T           | NPA(2.0)           | NE          |
| cudf.DataFrame          | T         | T                 | T         | T         | T           | T                  | T           |
| cudf.Series             | T         | T                 |           | T         |             | FF                 | NE          |
| cupy.ndarray            | T         | T                 | T         | T         | H           | T                  | T           |
| dlpack                  | CPA       | CPA               |           | CPA       | H           | FF                 | FF          |
| datatable.Frame         | T         | FF                |           | NPA       | F           | FF                 |             |
| datatable.Table         | T         | FF                |           | NPA       | F           | FF                 |             |
| modin.DataFrame         | NPA       | FF                |           | NPA       | T           | FF                 |             |
| modin.Series            | NPA       | FF                |           | NPA       | T           | FF                 |             |
| pyarrow.Table           | T         | F                 |           | NPA       | FF          | FF                 |             |
| pyarrow.dataset.Dataset | T         | F                 |           |           | F           | F                  |             |
| _\_array\_\_            | NPA       | F                 | NPA       | NPA       | H           | H                  |             |
| Others                  | SciCSR    | F                 |           | F         | H           | F                  |             |

## distributed data structures
WIP.

**Comment by trivialfis** [comment]
Related https://github.com/dmlc/xgboost/issues/8112 .

**Comment by thomasaarholt** [comment]
Great that this is being aggregated!

What does "full support" mean? And how does full-support relate to documentation? (I wrote #8112 :) )

**Comment by trivialfis** [comment]
Apologies for the ambiguity, `T` for predictor means XGBoost can consume the data directly, and in most of the cases (except for scenarios like np array having `object` or `float16` dtype) XGBoost doesn't need to make any extra data copy. While for `NPA`, the input is well supported, but we might need to transfer the data to numpy array via functions like `pd.DataFrame.values`, which may or may not make data copies, depending on pandas instead of XGBoost.

I will change the wording.
